src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -3810,4 +3810,50 @@ BOOST_AUTO_TEST_CASE(TransposedConvolutionalLayerOptionalParameterTest)
     delete decoder;
 }
 
+BOOST_AUTO_TEST_CASE(BatchNormWithMinBatchesTest)
+{
+  arma::mat input, output, result, runningMean, runningVar, delta;
+
+  input = arma::mat(8, 3);
+  input << 1 << 446 << 42 << arma::endr
+      << 2 << 16 << 63 << arma::endr
+      << 3 << 13 << 63 << arma::endr
+      << 4 << 21 << 21 << arma::endr
+      << 1 << 13 << 11 << arma::endr
+      << 32 << 45 << 42 << arma::endr
+      << 22 << 16 << 63 << arma::endr
+      << 32 << 13 << 42 << arma::endr;
+
+  // Output calculated using torch.nn.BatchNorm2d().
+  result = arma::mat(8, 3);
+  result << -0.4786 << 3.2634 << -0.1338 << arma::endr
+      << -0.4702 << -0.3525 << 0.0427 << arma::endr
+      << -0.4618 << -0.3777 << 0.0427 << arma::endr
+      << -0.4534 << -0.310433 << -0.310433 << arma::endr
+      << -1.5429 << -0.8486 << -0.9643 << arma::endr
+      << 0.2507 << 1.0029 << 0.8293 << arma::endr
+      << -0.32786 << -0.6750 << 2.0443 << arma::endr
+      << 0.2507 << -0.8486 << 0.8293 << arma::endr;
+
+  // Check correctness of batch normalization.
+  BatchNorm<> module1(2);
+  module1.Reset();
+  module1.Forward(input, output);
+  CheckMatrices(output, result, 1e-1);
+
+  // Check values for running mean and running variance.
+  runningMean = arma::mat(2, 1);
+  runningVar = arma::mat(2, 1);
+  runningMean << 5.8000 << 2.76667;
+  runningVar << 1543.7545 << 33.4879;
+
+  module1.TrainingMean().print();
+  module1.TrainingVariance().print();
```

Comment:
Forward pass is complete, There is an error inrunning variance. The values of mean (including running) and variance match however the value for running variance doesn't. So the output is correct but backward and gradient will be incorrect. Working on fixing this now.

Reply:
Fixed the error. The reference I was using didn't match with PyTorch's implementation completely and hence gave me errors so I am only using PyTorch as reference. [NoteBook Link](https://colab.research.google.com/drive/12Yqa6G9fqkLukUIG5s6denFnx_gfSnr5?usp=sharing). Now we can make the changes in backward and gradient. I have added tests for forward for both training and testing.

Reply:
@kartikdutt18 The notebook link isn't accessible, can you check it?

Reply:
Sorry about that, This [one](https://colab.research.google.com/drive/12Yqa6G9fqkLukUIG5s6denFnx_gfSnr5?usp=sharing) should work.

Reply:
Let me know once the PR is ready for review.

Reply:
Will do, Only backward function is left. After that it should be ready from my side.